### PR TITLE
feat(neptune): add engine version 1.2.0.0

### LINKED ISSUE #21877: neptune: support latest engine major version 1.2
### Describe the feature

AWS Neptune has released a new major version in 1.2 in https://docs.aws.amazon.com/neptune/latest/userguide/engine-releases-1.2.0.0.html which is not supported in [EngineVersion](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-neptune/lib/cluster.ts#L16) enum.

Additionally, parameters e.g https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-neptune/lib/parameter-group.ts#L84 seems to assume `family: 'neptune1'` without exposing a prop for setting that.

### Use Case

Use the latest MV 1.2 of Neptune

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.38.1

### Environment details (OS name and version, etc.)

macOs

### PR DESCRIPTION
**Pull request by humanzz**
- add engine version 1.2.0.0
- introduce a new enum for parameter group family
- update parameter groups to support specifying parameter group family

closes #21877

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by humanzz** [comment]
A few useful links for the review
- https://docs.aws.amazon.com/neptune/latest/userguide/engine-releases.html
- https://docs.aws.amazon.com/neptune/latest/userguide/engine-releases-1.2.0.0.html
- https://docs.aws.amazon.com/neptune/latest/userguide/iam-data-access-policies.html

**Review by TheRealAmazonKendra — DISMISSED** [review]
This is a very nice set of changes overall, but it does significantly more than what's in the PR title. Can you please split this PR up into PRs for each of the pieces of functionality you're adding? That way it will be accurately reflected in our change log and we can more thoroughly review each change on its own.

**Comment by humanzz** [comment]
@TheRealAmazonKendra Thing is, I think they're all related to the recently introduced engine version.
1. A new engine version enum introduced
2. grant method introduced as the new engine version supports the more granular actions
3. Parameter group family introduced as the new version requires a new family different from what was hardcoded

So, I think it'd be tough to split things in a way that would make sense. What do you think?

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Thing is, I think they're all related to the recently introduced engine version.
> 
> 1. A new engine version enum introduced
> 2. grant method introduced as the new engine version supports the more granular actions
> 3. Parameter group family introduced as the new version requires a new family different from what was hardcoded
> 
> So, I think it'd be tough to split things in a way that would make sense. What do you think?

Taking a look back at this, I think 1 and 3 make sense together in one PR, but 2 should be its own PR.

**Comment by humanzz** [comment]
> > @TheRealAmazonKendra Thing is, I think they're all related to the recently introduced engine version.
> > 
> > 1. A new engine version enum introduced
> > 2. grant method introduced as the new engine version supports the more granular actions
> > 3. Parameter group family introduced as the new version requires a new family different from what was hardcoded
> > 
> > So, I think it'd be tough to split things in a way that would make sense. What do you think?
> 
> Taking a look back at this, I think 1 and 3 make sense together in one PR, but 2 should be its own PR.

I've dropped the `grant` method and its tests.

One question about the separate PR for it, does that need creating a new issue, or can I reference https://github.com/aws/aws-cdk/issues/21877 instead

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 1]
```diff
@@ -362,7 +366,7 @@ export abstract class DatabaseClusterBase extends Resource implements IDatabaseC
 
   public grantConnect(grantee: iam.IGrantable): iam.Grant {
     if (this.enableIamAuthentication === false) {
-      throw new Error('Cannot grant connect when IAM authentication is disabled');
+      throw new Error('Cannot grant when IAM authentication is disabled');
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-neptune/lib/cluster.ts:369]: This error message doesn't quite make sense worded this way.

File: packages/@aws-cdk/aws-neptune/test/cluster.test.ts [hunk 2]
```diff
@@ -473,7 +473,7 @@ describe('DatabaseCluster', () => {
     });
   });
 
-  test('createGrant - creates IAM policy and enables IAM auth', () => {
+  test('grantConnect - enables IAM auth and grants neptune-db:* to the specified grantee', () => {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-neptune/test/cluster.test.ts:476]: I think this change should have gone in the other PR.

File: packages/@aws-cdk/aws-neptune/test/cluster.test.ts [hunk 3]
```diff
@@ -528,7 +528,7 @@ describe('DatabaseCluster', () => {
     });
   });
 
-  test('createGrant - throws if IAM auth disabled', () => {
+  test('grantConnect - throws if IAM auth disabled', () => {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-neptune/test/cluster.test.ts:531]: I think this change should have gone in the other PR.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 1 above)
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:369]: Actually, I think this wording change should go in the other PR anyway.

File: packages/@aws-cdk/aws-neptune/test/cluster.test.ts [hunk 4]
```diff
@@ -544,7 +544,7 @@ describe('DatabaseCluster', () => {
     });
 
     // THEN
-    expect(() => { cluster.grantConnect(role); }).toThrow(/Cannot grant connect when IAM authentication is disabled/);
+    expect(() => { cluster.grantConnect(role); }).toThrow(/Cannot grant when IAM authentication is disabled/);
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-neptune/test/cluster.test.ts:547]: I think this change should have gone in the other PR.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Just a few very minor things. Please see my comments inline.

File: packages/@aws-cdk/aws-neptune/test/integ.cluster-ev12.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
```
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-neptune/test/integ.cluster-ev12.ts]: You can see how that would look like in https://github.com/aws/aws-cdk/pull/21908/commits/7e9858c44ba916f7becbdcfeb1d79db41cfc3530

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-neptune/test/integ.cluster-ev12.ts]: I got that. Tbh, I was hoping that tests will be done in the requested fashion. In reality, observing the cloud formation stacks, I saw the 2 stacks get created first then they get destroyed, it was not one after the other.

I can revert that last commit.

Any other comments to address? Happy with integ tests file names and their usage of the new construct?

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 1 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts]: Dropped from here, and will address separately in the other PR

**Review by TheRealAmazonKendra — APPROVED** [review]
Looks great!

**Pull request merged**